versioned/storage/cache/src/main/java/org/projectnessie/versioned/storage/cache/CacheSizing.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.storage.cache;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import java.util.OptionalDouble;
+import java.util.OptionalInt;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface CacheSizing {
+
+  int DEFAULT_HEAP_SIZE_KEEP_FREE = 256;
+
+  OptionalInt fixedSizeInMB();
+
+  OptionalDouble fractionOfMaxHeapSize();
+
+  OptionalInt fractionMinSizeMb();
+
+  @Value.Default
+  default int heapSizeAdjustmentMB() {
+    return DEFAULT_HEAP_SIZE_KEEP_FREE;
+  }
+
+  default int calculateEffectiveSizeInMB() {
+    return calculateEffectiveSizeInMB(Runtime.getRuntime().maxMemory());
+  }
+
+  default int calculateEffectiveSizeInMB(long maxHeapInBytes) {
+    if (fractionOfMaxHeapSize().isPresent()) {
+      long fractionAsBytes = (long) (fractionOfMaxHeapSize().getAsDouble() * maxHeapInBytes);
+
+      long freeHeap = maxHeapInBytes - fractionAsBytes;
+      long minFree = heapSizeAdjustmentMB() * 1024L * 1024L;
+
+      long capacityInBytes = (minFree > freeHeap) ? maxHeapInBytes - minFree : fractionAsBytes;
+
+      if (fractionMinSizeMb().isPresent()) {
+        long fractionMin = (long) fractionMinSizeMb().getAsInt() * 1024 * 1024;
+        if (capacityInBytes < fractionMin) {
+          capacityInBytes = fractionMin;
+        }
+      }
+
+      return (int) (capacityInBytes / 1024L / 1024L);
+    }
+
+    return fixedSizeInMB().orElseThrow();
+  }
+
+  static Builder builder() {
+    return ImmutableCacheSizing.builder();
+  }
+
+  interface Builder {
+    @CanIgnoreReturnValue
+    Builder fixedSizeInMB(int fixedSizeInMB);
+
+    @CanIgnoreReturnValue
+    Builder fractionOfMaxHeapSize(double fractionOfMaxHeapSize);
+
+    @CanIgnoreReturnValue
+    Builder fractionMinSizeMb(int fractionMinSizeMb);
+
+    @CanIgnoreReturnValue
+    Builder heapSizeAdjustmentMB(int heapSizeAdjustmentMB);
+
+    CacheSizing build();
+  }
+
+  @Value.Check
+  default void check() {
+    checkState(
+        !(fixedSizeInMB().isPresent() && fractionOfMaxHeapSize().isPresent()),
+        "Cache sizing: must not specify both fractionOfMaxHeapSize and sizeInBytes, use either.");
```

Comment:
This looks a bit too restrictive to me. Could we make one of the configuration sub-sets take precedence instead of making them mutually exclusive? That would simplify upgrades with Helm, I hope. Currently the old fix size settings are often present in `application.properties`, which ties them to binary artifact deployment.

Reply:
Updated